src/Microsoft.IdentityModel.Tokens/ProvidersCache/HashCache.cs

```diff
@@ -0,0 +1,264 @@
+﻿//------------------------------------------------------------------------------
+//
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.CryptoProviderCacheOptions
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Microsoft.IdentityModel.Logging;
+
+namespace Microsoft.IdentityModel.Tokens
+{
+    internal class HashCache<TKey, TValue> : IProviderCache<TKey, TValue>
```

Comment:
>HashCache [](http://example.com/codeflow?start=19&length=9)

since this one removes from the front, should this be called FIFOCache or similar? Or we can stick with SimpleEvictionCache

Reply:
I welcome suggestions on the name, Hash clearly says it is hash-based but SimpleEvictionCache sounds good too.

Reply:
I like FIFOCache since it's clear on exactly what happens when the cache needs to evict items.

Reply:
I don't think Dictionary guarantees FIFO so this might not be a good choice. I think SimpleEvictionCache is more generic and applies to how it works.